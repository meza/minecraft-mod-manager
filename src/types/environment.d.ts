declare global {
  // eslint-disable-next-line no-unused-vars
  namespace NodeJS {
    // eslint-disable-next-line no-unused-vars
    interface ProcessEnv {
      CURSEFORGE_API_KEY: string,
      MODRINTH_API_KEY: string,
      POSTHOG_API_KEY: string,
      npm_package_version: string
    }
  }
}

export {};
