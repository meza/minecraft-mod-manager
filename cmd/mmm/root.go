package mmm

import (
	"fmt"
	initCmd "github.com/meza/minecraft-mod-manager/cmd/mmm/init"
	"github.com/meza/minecraft-mod-manager/cmd/mmm/version"
	"github.com/meza/minecraft-mod-manager/internal/constants"
	"github.com/meza/minecraft-mod-manager/internal/environment"
	"github.com/meza/minecraft-mod-manager/internal/i18n"
	"github.com/meza/minecraft-mod-manager/internal/tui"
	"github.com/spf13/cobra"
	"golang.org/x/term"
	"os"
	"strings"
)

func Command() *cobra.Command {
	rootCmd := &cobra.Command{
		Use:     constants.COMMAND_NAME,
		Short:   i18n.T("app.description"),
		Version: environment.AppVersion(),
		RunE:    runRoot,
	}

	rootCmd.SetVersionTemplate("{{.Version}}\n")
	rootCmd.AddCommand(initCmd.Command())
	rootCmd.AddCommand(version.Command())

	translateDefaultHelpFacilities(rootCmd)
	fixFlagUsageAlignment(rootCmd)

	return rootCmd
}

func translateDefaultHelpFacilities(rootCmd *cobra.Command) {
	subcommands := rootCmd.Commands()
	allCommands := make([]*cobra.Command, 0, len(subcommands)+1)
	allCommands = append(allCommands, rootCmd)
	allCommands = append(allCommands, subcommands...)

	for _, cmd := range allCommands {
		cmd.InitDefaultHelpFlag()
		flags := cmd.Flags()
		flags.Lookup("help").Usage = i18n.T("cmd.help.template", i18n.Tvars{
			Data: &i18n.TData{"command": cmd.Name()},
		})
	}

	rootCmd.InitDefaultHelpCmd()
	helpCmd, _, e := rootCmd.Find([]string{"help"})

	if e == nil {
		helpCmd.Short = i18n.T("cmd.help.usage.short")
		helpCmd.Long = i18n.T("cmd.help.usage.long", i18n.Tvars{
			Data: &i18n.TData{"appName": rootCmd.Name()},
		})
		helpCmd.Run = func(c *cobra.Command, args []string) {
			cmd, _, e := c.Root().Find(args)
			if cmd == nil || e != nil {
				c.PrintErrln(i18n.T("cmd.help.error", i18n.Tvars{
					Data: &i18n.TData{"topic": fmt.Sprintf("%#q", args)},
				}) + "\n")
				cobra.CheckErr(c.Root().Usage())
			} else {
				cmd.InitDefaultHelpFlag()    // make possible 'help' flag to be shown
				cmd.InitDefaultVersionFlag() // make possible 'version' flag to be shown
				cobra.CheckErr(cmd.Help())
			}
		}
	}
}

func fixFlagUsageAlignment(rootCmd *cobra.Command) {
	width, _, _ := term.GetSize(int(os.Stdout.Fd()))
	usageTemplate := rootCmd.UsageTemplate()
	usageTemplate = strings.ReplaceAll(usageTemplate, ".FlagUsages", fmt.Sprintf(".FlagUsagesWrapped %d", width))
	rootCmd.SetUsageTemplate(usageTemplate)
}

func runRoot(_ *cobra.Command, _ []string) error {
	return tui.RunApp([]tui.Component{
		initCmd.Component(),
	})
}

func Execute() error {
	return Command().Execute()
}
