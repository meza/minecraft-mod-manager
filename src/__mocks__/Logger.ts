import { vi } from 'vitest';
import { Logger as Original } from '../lib/Logger.js';

export const Logger = vi.importMock('../lib/Logger.js') as unknown as Original;
