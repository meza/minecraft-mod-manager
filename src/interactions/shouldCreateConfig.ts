import { confirm } from '@inquirer/prompts';

export const shouldCreateConfig = async (configLocation: string): Promise<boolean> => {
  return confirm({
    default: true,
    message: `The config file: (${configLocation}) does not exist. Should we create it?`
  });
};
