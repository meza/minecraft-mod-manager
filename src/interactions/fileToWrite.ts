import fs from 'node:fs/promises';
import path from 'node:path';
import { input } from '@inquirer/prompts';
import chalk from 'chalk';
import { Logger } from '../lib/Logger.js';
import { fileExists } from '../lib/config.js';
import { DefaultOptions } from '../mmm.js';

const isFileWritable = async (filePath: string) => {
  let absolutePath = path.resolve(filePath);

  if (!(await fileExists(filePath))) {
    absolutePath = path.resolve(path.dirname(filePath));
  }

  const permissions = fs.constants.W_OK;

  try {
    await fs.access(absolutePath, permissions);
    return true;
  } catch {
    return false;
  }
};

export const fileToWrite = async (filePath: string, options: DefaultOptions, logger: Logger) => {
  logger.debug(`Checking if ${filePath} is writable`);
  const result = await isFileWritable(filePath);

  if (result === true) {
    return filePath;
  }

  if (options.quiet) {
    logger.error(`${filePath} is not writable. Aborting.`, 1);
  }

  const answer = await input({
    message: chalk.bgRed(chalk.whiteBright(`${filePath} is not writable, please choose another one`)),
    default: filePath,
    validate: async (input: string) => {
      if (await isFileWritable(input)) {
        return true;
      }
      return chalk.bgRed(chalk.whiteBright(`${input} is not writable, please choose another one`));
    }
  });

  return answer;
};
