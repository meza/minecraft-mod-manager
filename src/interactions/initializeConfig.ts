import path from 'node:path';
import { checkbox, input, select } from '@inquirer/prompts';
import { IncorrectMinecraftVersionException } from '../errors/IncorrectMinecraftVersionException.js';
import { Logger } from '../lib/Logger.js';
import { fileExists, writeConfigFile } from '../lib/config.js';
import { verifyMinecraftVersion } from '../lib/minecraftVersionVerifier.js';
import { Loader, ModsJson, ReleaseType } from '../lib/modlist.types.js';
import { DefaultOptions } from '../mmm.js';
import { configFile } from './configFileOverwrite.js';
import { getLatestMinecraftVersion } from './getLatestMinecraftVersion.js';

export interface InitializeOptions extends DefaultOptions {
  loader?: Loader;
  gameVersion?: string;
  allowVersionFallback?: boolean;
  defaultAllowedReleaseTypes?: string;
  modsFolder?: string;
}

interface AnswersInternal {
  loader?: Loader;
  gameVersion?: string;
  allowVersionFallback?: boolean;
  defaultAllowedReleaseTypes?: string[];
  modsFolder?: string;
  config: string;
}

const validateGameVersion = async (input: string): Promise<boolean | string> => {
  if (await verifyMinecraftVersion(input)) {
    return true;
  }
  return 'The game version is invalid. Please enter a valid game version';
};

const mergeOptions = (options: InitializeOptions, iq: AnswersInternal) => {
  return {
    loader: iq.loader || options.loader,
    gameVersion: iq.gameVersion || options.gameVersion,
    defaultAllowedReleaseTypes:
      iq.defaultAllowedReleaseTypes || options.defaultAllowedReleaseTypes?.replace(/\s/g, '').split(','),
    modsFolder: iq.modsFolder || options.modsFolder,
    mods: []
  };
};

const validateModsFolder = async (input: string, cwd: string) => {
  let dir = path.resolve(cwd, input);
  if (path.isAbsolute(input)) {
    dir = input;
  }

  if (!(await fileExists(dir))) {
    return `The folder: ${dir} does not exist. Please enter a valid one and try again.`;
  }
  return true;
};

const validateInput = async (options: InitializeOptions, cwd: string) => {
  /**
   * @todo Handle cli option validation better for the init function
   *
   * Currently the negative case is just throwing errors. It would be nice
   * to properly communicate the errors and offer solutions.
   */
  if (options.gameVersion) {
    if (!(await verifyMinecraftVersion(options.gameVersion))) {
      throw new IncorrectMinecraftVersionException(options.gameVersion);
    }
  }
  if (options.modsFolder) {
    const result = await validateModsFolder(options.modsFolder, cwd);
    if (result !== true) {
      throw new Error(result);
    }
  }
};

export const initializeConfig = async (options: InitializeOptions, cwd: string, logger: Logger): Promise<ModsJson> => {
  await validateInput(options, cwd);

  options.config = await configFile(options, cwd);

  const latestMinercraftDefault = await getLatestMinecraftVersion(options, logger);

  const answers: AnswersInternal = {
    config: options.config
  };

  if (!options.loader) {
    answers.loader = await select({
      message: 'Which loader would you like to use?',
      choices: Object.values(Loader)
    });
  }

  if (!options.gameVersion) {
    answers.gameVersion = await input({
      default: latestMinercraftDefault,
      message: 'What exact Minecraft version are you using? (eg: 1.18.2, 1.19, 1.19.1)',
      validate: validateGameVersion
    });
  }

  if (!options.defaultAllowedReleaseTypes) {
    answers.defaultAllowedReleaseTypes = await checkbox({
      message: 'Which types of releases would you like to consider to download?',
      choices: getReleaseTypeChoices([ReleaseType.RELEASE, ReleaseType.BETA])
    });
  }

  if (!options.modsFolder) {
    answers.modsFolder = await input({
      default: './mods',
      message: `where is your mods folder? (full or relative path from ${cwd}):`,
      validate: async (input: string) => {
        return validateModsFolder(input, cwd);
      }
    });
  }

  const answersOut = mergeOptions(options, answers) as ModsJson;

  await writeConfigFile(answersOut, options, logger);

  return answersOut as ModsJson;
};

interface Choice {
  name: string;
  value: string;
  checked?: boolean;
}

const getReleaseTypeChoices = (defaults: ReleaseType[]): Choice[] => {
  return Object.values(ReleaseType).map((value) => {
    const choice: Choice = {
      name: value,
      value: value
    };

    if (defaults.includes(value)) {
      choice.checked = true;
    }

    return choice;
  });
};
