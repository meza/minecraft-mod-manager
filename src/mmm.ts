#!/usr/bin/env node
import { Command } from 'commander';
import 'dotenv/config';
import { add } from './actions/add.js';
import { install } from './actions/install.js';
import { list } from './actions/list.js';
import { testGameVersion } from './actions/testGameVersion.js';
import { update } from './actions/update.js';
import { helpUrl } from './env.js';
import { initializeConfig } from './interactions/initializeConfig.js';
import { Logger } from './lib/Logger.js';
import { Loader, Platform, ReleaseType } from './lib/modlist.types.js';
import { Telemetry } from './telemetry/telemetry.js';
import { version } from './version.js';
import { changeGameVersion } from './actions/change.js';
import { scan } from './actions/scan.js';
import { prune } from './actions/prune.js';
import { removeAction } from './actions/remove.js';

performance.mark('start');

export const APP_NAME = 'Minecraft Mod Manager';
export const APP_DESCRIPTION = 'Manages mods from Modrinth and Curseforge';
export const DEFAULT_CONFIG_LOCATION = './modlist.json';
const cwd = process.cwd();

export enum EXIT_CODE {
  SUCCESS = 0,
  GENERAL_ERROR = 1,
  SUPPLEMENTARY_ERROR = 2
}

export interface DefaultOptions {
  config: string;
  debug?: boolean;
  quiet?: boolean;
}

export const program = new Command();

export const logger: Logger = new Logger(program);
export const telemetry: Telemetry = new Telemetry();

export const stop = async (): Promise<never> => {
  await telemetry.flush();
  // eslint-disable-next-line no-process-exit
  process.exit(-1);
};

const commands = [];

program.name(APP_NAME).version(version).description(APP_DESCRIPTION);
program.addHelpText('after', '\nFor more information, visit: ' + helpUrl);

program.on('option:quiet', () => {
  logger.flagQuiet();
});

program.on('option:debug', () => {
  logger.flagDebug();
});

commands.push(
  program.command('list')
    .action(async (_options, cmd) => {
      await list(cmd.optsWithGlobals(), logger);
    })
    .aliases(['l', 'ls'])
);

commands.push(
  program.command('install')
    .action(async (_options, cmd) => {
      await install(cmd.optsWithGlobals(), logger);
    })
    .aliases(['i'])
);

commands.push(
  program.command('update')
    .action(async (_options, cmd) => {
      await update(cmd.optsWithGlobals(), logger);
    })
    .aliases(['u'])
);

commands.push(
  program.command('add')
    .argument('<type>', 'curseforge or modrinth')
    .argument('<id>', 'Curseforge or Modrinth Project Id')
    .option('-v, --version <version>', 'The version of the mod to add. If not specified, the latest version will be used')
    .option('-f, --allow-version-fallback', 'Should we try to download the mod for previous Minecraft versions if they do not exists for your Minecraft Version?', false)
    .action(async (type: Platform, id: string, _options, cmd) => {
      await add(type, id, cmd.optsWithGlobals(), logger);
    })
    .aliases(['a'])
);

commands.push(
  program.command('init')
    .option('-l, --loader <loader>', `Which loader would you like to use? ${Object.values(Loader).join(', ')}`)
    .option('-g, --game-version <gameVersion>', 'What exact Minecraft version are you using? (eg: 1.18.2, 1.19, 1.19.1)')
    .option('-r, --default-allowed-release-types <defaultAllowedReleaseTypes>',
      `Which types of releases would you like to consider to download? ${Object.values(ReleaseType).join(', ')} - comma separated list`)
    .option('-m, --mods-folder <modsFolder>', `where is your mods folder? (full or relative path from ${cwd})`)
    .action(async (_options, cmd) => {
      await initializeConfig(cmd.optsWithGlobals(), cwd, logger);
    })
);

commands.push(
  program.command('test')
    .argument('[game_version]', 'The Minecraft version to test', 'latest')
    .aliases(['t'])
    .action(async (gameVersion: string, _options, cmd) => {
      await testGameVersion(gameVersion, cmd.optsWithGlobals(), logger);
    })
);

commands.push(
  program.command('change')
    .option('-f, --force', 'Force the change of the game version. Deletes all the mods and attempts to install with the given game version. Use at your own risk', false)
    .argument('[game_version]', 'The Minecraft version to change to', 'latest')
    .action(async (gameVersion: string, _options, cmd) => {
      await changeGameVersion(gameVersion, cmd.optsWithGlobals(), logger);
    })
);

commands.push(
  program.command('scan')
    .description('Scans the mod directory and attempts to find the mods on the supported mod platforms.')
    .option('-p, --prefer <platform>', `Which platform do you prefer to use? ${Object.values(Platform).join(', ')}`, Platform.MODRINTH)
    .option('-a, --add', 'Add the mods to the modlist.json file', false)
    .action(async (_options, cmd) => {
      await scan(cmd.optsWithGlobals(), logger);
    })
);

commands.push(
  program.command('prune')
    .description('Prunes the mod directory from all the unmanaged files.')
    .option('-f, --force', 'Delete the files without asking', false)
    .action(async (_options, cmd) => {
      await prune(cmd.optsWithGlobals(), logger);
    })
);

commands.push(
  program.command('remove')
    .description('Removes one or more mods from both the config and the filesystem.')
    .option('-n, --dry-run', 'Print out the files/mods that would have been removed', false)
    .argument('<mods...>', 'A list of the mod(s) to remove. e.g: mmm remove mod1 mod2 "mod with space in its name"')
    .action(async (mods: string[], _options, cmd) => {
      await removeAction(mods, cmd.optsWithGlobals(), logger);
    })
);

program.option('-c, --config <MODLIST_JSON>', 'An alternative JSON file containing the configuration', DEFAULT_CONFIG_LOCATION);
program.option('-q, --quiet', 'Suppress all output', false);
program.option('-d, --debug', 'Enable debug messages', false);
