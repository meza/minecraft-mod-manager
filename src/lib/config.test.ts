import { beforeEach, describe, expect, it, vi } from 'vitest';
import fs from 'node:fs/promises';
import { chance } from 'jest-chance';
import {
  ensureConfiguration,
  fileExists, getModsFolder,
  initializeConfigFile, ModsJsonSchema,
  readConfigFile,
  readLockFile,
  writeConfigFile,
  writeLockFile
} from './config.js';
import { Loader, ModInstall, ModsJson, Platform, ReleaseType } from './modlist.types.js';
import path from 'node:path';
import { generateModInstall } from '../../test/modInstallGenerator.js';
import { generateModsJson } from '../../test/modlistGenerator.js';
import { ConfigFileNotFoundException } from '../errors/ConfigFileNotFoundException.js';
import { initializeConfig } from '../interactions/initializeConfig.js';
import * as process from 'process';
import { shouldCreateConfig } from '../interactions/shouldCreateConfig.js';
import { Logger } from './Logger.js';
import { fileToWrite } from '../interactions/fileToWrite.js';
import { DefaultOptions } from '../mmm.js';

vi.mock('../interactions/shouldCreateConfig.js');
vi.mock('../interactions/initializeConfig.js');
vi.mock('node:fs/promises');
vi.mock('../interactions/fileToWrite.js');
vi.mock('../lib/Logger.js');

interface LocalTestContext {
  options: DefaultOptions;
}

describe('The config library', () => {
  let logger: Logger;
  beforeEach<LocalTestContext>((context) => {
    vi.resetAllMocks();
    vi.unstubAllGlobals();
    logger = new Logger({} as never);
    context.options = {
      config: 'config.json',
      debug: false,
      quiet: false
    };
    vi.mocked(fileToWrite).mockImplementation(async (file: string) => {
      return file;
    });
    vi.mocked(logger.error).mockImplementation((message) => {
      throw new Error(message);
    });
  });

  it('can determine that a file exists', async () => {
    vi.mocked(fs.access).mockResolvedValueOnce();
    expect(await fileExists(chance.word())).toBeTruthy();
  });

  it('can determine that a file does not exist', async () => {
    vi.mocked(fs.access).mockRejectedValueOnce(new Error());
    expect(await fileExists(chance.word())).toBeFalsy();
  });

  it<LocalTestContext>('can write the config file', async ({ options }) => {
    const config = {
      something: 'value'
    } as unknown as ModsJson;

    await writeConfigFile(config, options, logger);

    expect(vi.mocked(fileToWrite)).toHaveBeenCalledWith(path.resolve(options.config), options, logger);

    expect(vi.mocked(fs.writeFile)).toHaveBeenCalledWith(
      path.resolve(options.config),
      '{\n'
      + '  "something": "value"\n'
      + '}'
    );
  });

  it<LocalTestContext>('can write the lock file', async ({ options }) => {
    const config = [{
      something: 'value'
    }] as unknown as ModInstall[];
    options.config = chance.word();
    const expectedLockFilePath = path.resolve(`${options.config}-lock.json`);

    await writeLockFile(config, options, logger);
    expect(vi.mocked(fileToWrite)).toHaveBeenCalledWith(expectedLockFilePath, options, logger);

    expect(vi.mocked(fs.writeFile)).toHaveBeenCalledWith(
      expectedLockFilePath,
      '[\n'
      + '  {\n'
      + '    "something": "value"\n'
      + '  }\n'
      + ']'
    );
  });

  it<LocalTestContext>('can read the lock file when it exists', async ({ options }) => {
    options.config = 'config.json';
    const lockfileName = 'config-lock.json';

    const randomModInstall = generateModInstall();

    const lockfileContents = JSON.stringify([randomModInstall.generated], null, 2);

    // Make file exist
    vi.mocked(fs.access).mockResolvedValueOnce();
    // Return file contents
    vi.mocked(fs.readFile).mockResolvedValueOnce(lockfileContents);
    const actualOutput = await readLockFile(options, logger);

    expect(actualOutput).toEqual([randomModInstall.expected]);

    expect(vi.mocked(fs.readFile)).toHaveBeenCalledWith(
      path.resolve(lockfileName),
      { encoding: 'utf8' }
    );

  });

  it<LocalTestContext>('can returns an empty array and creates the file when the lock file does not exist', async ({ options }) => {
    options.config = 'config.json';

    // Make file exist
    vi.mocked(fs.access).mockRejectedValueOnce(new Error());

    const actualOutput = await readLockFile(options, logger);

    expect(actualOutput).toEqual([]);

    expect(vi.mocked(fs.readFile)).not.toHaveBeenCalled();
    expect(vi.mocked(fs.writeFile)).toHaveBeenCalledWith(
      path.resolve('config-lock.json'),
      '[]'
    );

  });

  it<LocalTestContext>('can read from the config file when it exists', async ({ options }) => {
    options.config = 'config.json';

    // File exists
    vi.mocked(fs.access).mockResolvedValueOnce();

    const randomModsJson = generateModsJson();
    const fileContents = JSON.stringify(randomModsJson.generated, null, 2);

    // Return config file contents
    vi.mocked(fs.readFile).mockResolvedValueOnce(fileContents);

    const actualOutput = await readConfigFile(options.config);

    expect(actualOutput).toEqual(randomModsJson.expected);
  });

  it('throws an error when the config file does not exist', async () => {
    const configName = path.resolve('config.json');

    // File does not exist
    vi.mocked(fs.access).mockRejectedValueOnce(new Error());

    await expect(readConfigFile(configName)).rejects.toThrow(new ConfigFileNotFoundException(configName));
  });

  it('throws an error when there are misconfigurations', async () => {
    const configName = 'config.json';
    const invalidModsJson = {
      loader: 'invalid_loader',
      gameVersion: '1.16.5',
      defaultAllowedReleaseTypes: ['invalid_release_type'],
      modsFolder: 'mods',
      mods: [
        {
          id: 'mod1',
          platform: 'invalid_platform',
          version: '1.0.0',
          allowVersionFallback: true,
          allowedReleaseTypes: ['invalid_release_type']
        }
      ]
    };

    const fileContents = JSON.stringify(invalidModsJson, null, 2);
    vi.mocked(fs.access).mockResolvedValueOnce();
    // Return config file contents
    vi.mocked(fs.readFile).mockResolvedValueOnce(fileContents);
    await expect(ensureConfiguration(configName, logger)).rejects.toThrowErrorMatchingSnapshot();
  });

  it('can initialize a new config file', async () => {
    const configName = 'config.json';

    const randomConfig = generateModsJson().generated;

    vi.mocked(initializeConfig).mockResolvedValueOnce(randomConfig);
    const actual = await initializeConfigFile(configName, logger);

    expect(actual).toEqual(randomConfig);
    expect(vi.mocked(initializeConfig)).toHaveBeenCalledWith({ config: configName }, process.cwd(), logger);

  });

  describe('when ensuring that a config exists', () => {
    describe('and in interactive mode', () => {
      it('creates a new one if there is no existing one', async () => {
        const configName = 'config.json';
        const randomConfig = generateModsJson().generated;

        //File does not exist
        vi.mocked(fs.access).mockRejectedValueOnce(new Error());
        vi.mocked(shouldCreateConfig).mockResolvedValueOnce(true);
        vi.mocked(initializeConfig).mockResolvedValueOnce(randomConfig);

        const actual = await ensureConfiguration(configName, logger);

        expect(actual).toEqual(randomConfig);
        expect(initializeConfig).toHaveBeenCalledOnce();
      });

      it('throws an error if the user chose to not create one', async () => {
        const configName = 'config.json';

        //File does not exist
        vi.mocked(fs.access).mockRejectedValueOnce(new Error());
        vi.mocked(shouldCreateConfig).mockResolvedValueOnce(false);

        await expect(ensureConfiguration(configName, logger)).rejects.toThrow(ConfigFileNotFoundException);

        expect(initializeConfig).not.toHaveBeenCalled();
      });
    });

    describe('and in non-interactive mode', () => {
      it('throws an error', async () => {
        const configName = 'config.json';
        const quiet = true;

        //File does not exist
        vi.mocked(fs.access).mockRejectedValueOnce(new Error());

        await expect(ensureConfiguration(configName, logger, quiet)).rejects.toThrow(ConfigFileNotFoundException);
      });
    });

    it('returns the existing one if there is one', async () => {
      const configName = 'config.json';

      // File exists
      vi.mocked(fs.access).mockResolvedValueOnce();

      const randomModsJson = generateModsJson();
      const fileContents = JSON.stringify(randomModsJson.generated, null, 2);

      // Return config file contents
      vi.mocked(fs.readFile).mockResolvedValueOnce(fileContents);

      const actualOutput = await ensureConfiguration(configName, logger);

      expect(actualOutput).toEqual(randomModsJson.expected);
    });
  });

  it('can resolve a relative mod folder', () => {
    const randomModsJson = generateModsJson().generated;
    const configPath = path.resolve('/some-path/config.json');
    const expected = path.resolve('/some-path/mods');
    randomModsJson.modsFolder = 'mods';
    const actual = getModsFolder(configPath, randomModsJson);

    expect(actual).toEqual(expected);
  });

  it('can resolve an absolute mod folder', () => {
    const randomModsJson = generateModsJson().generated;
    const configPath = '/some-path/config.json';
    const modsFolder = '/my-ultimate-mods';
    randomModsJson.modsFolder = modsFolder;
    const expected = '/my-ultimate-mods';
    const actual = getModsFolder(configPath, randomModsJson);

    expect(actual).toEqual(expected);
  });

  it('should validate a correct ModsJson object', () => {
    const validModsJson: ModsJson = {
      loader: Loader.FORGE,
      gameVersion: '1.16.5',
      defaultAllowedReleaseTypes: [ReleaseType.RELEASE, ReleaseType.BETA],
      modsFolder: 'mods',
      mods: [
        {
          id: 'mod1',
          type: Platform.CURSEFORGE,
          name: 'Mod 1',
          version: '1.0.0',
          allowVersionFallback: true,
          allowedReleaseTypes: [ReleaseType.RELEASE]
        }
      ]
    };

    const result = ModsJsonSchema.safeParse(validModsJson);
    expect(result.success).toBe(true);
  });

  it('should invalidate an incorrect ModsJson object', () => {
    const invalidModsJson = {
      loader: 'invalid_loader',
      gameVersion: '1.16.5',
      defaultAllowedReleaseTypes: ['invalid_release_type'],
      modsFolder: 'mods',
      mods: [
        {
          id: 'mod1',
          platform: 'invalid_platform',
          version: '1.0.0',
          allowVersionFallback: true,
          allowedReleaseTypes: ['invalid_release_type']
        }
      ]
    };

    const result = ModsJsonSchema.safeParse(invalidModsJson);
    expect(result.success).toBe(false);
  });
});
