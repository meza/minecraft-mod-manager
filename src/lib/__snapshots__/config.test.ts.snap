// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The config library > throws an error when there are misconfigurations 1`] = `[Error: There is a problem with the configuration file, please check!]`;
